tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
   return true;
 }
 
+//===----------------------------------------------------------------------===//
+// mhlo.Einsum conversion patterns.
+//===----------------------------------------------------------------------===//
+
+// Looks through a set of dimension that has been marked as Reduction axes
+// if it is found within the set, then we set it as "reduction", otherwise
+// we can label it as "parallel"
+SmallVector<StringRef, 3> GetEinsumLoopsAttrs(
+    const llvm::SmallSetVector<StringRef, 4>& input_ind,
+    const llvm::SmallSetVector<StringRef, 4>& reduction_dims) {
+  SmallVector<StringRef, 3> res;
+  for (const StringRef& dim : input_ind) {
+    if (!reduction_dims.contains(dim)) {
+      res.push_back(getParallelIteratorTypeName());
+    } else {
+      res.push_back(getReductionIteratorTypeName());
+    }
+  }
+  return res;
+}
+
+// Check for where each dimension of the output operand came from
+// If that dimension is dynamically sized then we push it to dyn_size vector
+SmallVector<Value, 2> ExtractDynamicEinsumSizes(
+    OpBuilder& b, const Location& loc, const Value& lhs, const Value& rhs,
+    const SmallVector<std::string>& lhs_loop_vec,
+    const SmallVector<std::string>& rhs_loop_vec,
+    const SmallVector<std::string>& output_loop_vec) {
+  std::vector<Value> tensor_vec = {lhs, rhs};
+  std::vector<ArrayRef<int64_t>> tensor_shape_vec;
+  for (auto tensor_val : tensor_vec) {
+    auto tensor_current_type =
+        tensor_val.getType().dyn_cast<RankedTensorType>();
+    if (!tensor_current_type) return {};
+    tensor_shape_vec.push_back(tensor_current_type.getShape());
+  }
```

Comment:
As below explanation, let's delete these. For RankedTensorType check, we should move it to the beginning of the pattern.